src/java.base/share/classes/java/security/SignatureSpi.java

```diff
@@ -392,19 +392,16 @@ protected void engineSetParameter(AlgorithmParameterSpec params)
      * <p>This method is overridden by providers to return the parameters
      * used with this signature engine.
      *
-     * <p> If this signature engine has been initialized with parameters
-     * (by calling {@link #engineSetParameter(AlgorithmParameterSpec)} or
-     * {@link #engineSetParameter(String, Object)}) and the underlying signature
-     * implementation supports returning the parameters as
-     * {@code AlgorithmParameters}, this method returns the same parameters.
-     * If the parameters were not set, this method may return a combination
-     * of default and randomly generated parameter values if the
-     * underlying signature implementation supports it and can successfully
-     * generate them. Otherwise, {@code null} is returned.
+     * <p>The returned parameters may be the same that were used to initialize
+     * this signature, or may contain additional default or random parameter
+     * values used by the underlying signature scheme. If the required
+     * parameters were not supplied and can be generated by the signature,
+     * the generated parameters will be returned. Otherwise, {@code null}
```

Comment:
Minor wording nit - change "will be" to "are" to be consistent with other wording which uses the present tense. Same comment applies to `Signature.getParameters`.

Also, do we no longer need to mention the part "and the underlying signature implementation supports returning the parameters as {@code AlgorithmParameters}"?

Reply:
Fixed the nit. Thanks~
As for the part about returning the parameters as `{@code AlgorithmParameters}`, it should be covered by current sentence, i.e. `can be generated by the signature`. Perhaps we don't have to spell out all possible causes?

Reply:
That sentence is specifically if the caller did not specify parameters though. The previous wording is if the caller did specify parameters (as say an `AlgorithmParameterSpec`) and the implementation cannot return them as `AlgorithmParameters`. So I think it needs to be mentioned as it is kind of surprising if you specify parameters to `init` but `getParameters` returns null.

Reply:
Hmm, would it fall under the "Otherwise, null is returned"? If not, perhaps we can add back the part about returning AlgorithmParameters as below:

**If the underlying signature implementation supports returning the parameters as {@code AlgorithmParameters},** the returned parameters may be the same that were used to initialize
 this signature, or may contain additional default or random parameter values used by the underlying signature scheme. If the required parameters were not supplied and can be generated by the signature, the generated parameters are returned. Otherwise, {@code null} is returned.

Reply:
Yes, I agree that it's a special case as the convention is to define the parameter spec class based on the ASN.1 structure. I like your suggestion better also. Will update using it.